### PR TITLE
Initial implementation for JsonBindingService

### PR DESCRIPTION
**Pull request by kaikreuzer**
Ported over from https://github.com/eclipse/smarthome/pull/5856

Also-By: Flavio Costa <flavio.costa.br@gmail.com>
Signed-off-by: Kai Kreuzer <kai@openhab.org>

### CODE REVIEW

**Comment by hakan42** [comment]
Many binding consuming JSON data from upstream API servers import and use the annotation com.google.gson.annotations.SerializedName to really clamp down the attribute naming.

How would this work if the specific json binding is abstracted away?

**Comment by kaikreuzer** [comment]
@flaviocosta-net Can you answer @hakan42's question?

**Comment by maggu2810** [comment]
I am missing an answer of this question for months, too.

See: https://github.com/eclipse/smarthome/issues/5884#issuecomment-408016900
(starting some comments above: https://github.com/eclipse/smarthome/issues/5884#issuecomment-406583949)

**Comment by maggu2810** [comment]
Correct me if I am wrong, but if we want to use JSON in the code base, we will develop and compile against `org.openhab.core.io.json` instead of an specific implementation (e.g. `org.openhab.core.io.json.gson`).
So, we need to create an annotation in the `org.openhab.core.io.json` bundle ourself.

With that in mind, I don't care if the annotation is named `JsonbProperty` or different as long as the name is not too exotic. :wink: 
Regardless if the name match to the JSON-B used one or not, this is just a very simple refactor if you switch some time...

**Comment by hakan42** [comment]
So we would first rely on an org.openhab....JsonProperty first and then at some future time just change the import at the top of the class? I must admit that I am partial to the SerializedName annotation, but this is my day-in-and-day-out work with Spring which trained my mind into a specific direction 😁 

The important part is really that we can decouple the class property names in the java class from the names in the serialized data stream.

File: bundles/org.openhab.core.io.json.gson/NOTICE [hunk 1]
```diff
@@ -0,0 +1,19 @@
+This content is produced and maintained by the Eclipse SmartHome project.
+
+* Project home: https://eclipse.org/smarthome/
```
Comment (wborn) [inline bundles/org.openhab.core.io.json.gson/NOTICE]: Please update this to an openHAB NOTICE file.

File: bundles/org.openhab.core.io.json/NOTICE [hunk 2]
```diff
@@ -0,0 +1,19 @@
+This content is produced and maintained by the Eclipse SmartHome project.
+
+* Project home: https://eclipse.org/smarthome/
```
Comment (wborn) [inline bundles/org.openhab.core.io.json/NOTICE]: Please update this to an openHAB NOTICE file.

**Comment by davidgraeff** [comment]
Is that still being developed?

**Comment by kaikreuzer** [comment]
I think this is waiting for a decision on https://github.com/openhab/openhab-core/pull/611#issuecomment-469041137...

**Comment by maggu2810** [comment]
Using official standards and to develop for APIs instead of specific implementation should always be preferred.
But #515 continues the specific [ESH PR](https://github.com/eclipse/smarthome/pull/5856) that has initial created on "Jul 4, 2018".

I don't know if the time that has been invested and needs to be further invested (development - review - cycles) makes sense.
If the PR is ready and just needs to be accepted, no need to turn it away.

So, how far are we from using a "official standard"?

**Comment by davidgraeff** [comment]
javax.json is the official way :D But everyone is using jackson or gson out there, without abstraction layer. I don't understand why OH needs one.

**Comment by kaikreuzer** [comment]
> If the PR is ready and just needs to be accepted, no need to turn it away.

Well, it is ready, but it causes (functional) conflicts with #611, which are not easy to overcome.

If @flaviocosta-net is fine with it, I'd therefore suggest to close this PR despite all the effort that already went into this feature.

**Pull request closed without merging**